### PR TITLE
resource/aws_route53_resolver_dnssec_config - new resource

### PR DESCRIPTION
**Pull request by shuheiktgw**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes https://github.com/hashicorp/terraform-provider-aws/issues/16837

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* **New Resource:** `aws_route53_resolver_dnssec_config`
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSRoute53ResolverDnssecConfig_*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverDnssecConfig_* -timeout 120m
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_basic
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_basic
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_disappear
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_disappear
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_disappear_VPC
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_disappear_VPC
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_basic
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_disappear_VPC
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_disappear
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_disappear_VPC (159.43s)
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_basic (287.57s)
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_disappear (289.48s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	291.487s
```

Thank you for your review! 👍

### CODE REVIEW

File: aws/resource_aws_route53_resolver_dnssec_config.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+	req := &route53resolver.UpdateResolverDnssecConfigInput{
+		ResourceId: aws.String(d.Get("resource_id").(string)),
+		Validation: aws.String(route53resolver.ValidationEnable),
+	}
+
+	log.Printf("[DEBUG] Creating Route53 Resolver DNSSEC config: %#v", req)
+	resp, err := conn.UpdateResolverDnssecConfig(req)
+	if err != nil {
+		return fmt.Errorf("error creating Route53 Resolver DNSSEC config: %s", err)
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_dnssec_config.go:68]: [Prefer `%w`](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/error-handling.md#wrap-errors).

```suggestion
		return fmt.Errorf("error creating Route53 Resolver DNSSEC config: %w", err)
```

File: aws/resource_aws_route53_resolver_dnssec_config.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	return resourceAwsRoute53ResolverDnssecConfigRead(d, meta)
+}
+
+func resourceAwsRoute53ResolverDnssecConfigRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).route53resolverconn
+	ec2Conn := meta.(*AWSClient).ec2conn
+
+	vpc, err := vpcDescribe(ec2Conn, d.Id())
+	if err != nil {
+		return fmt.Errorf("error getting VPC associated with Route53 Resolver DNSSEC config (%s): %s", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_dnssec_config.go:89]: [Prefer `%w`](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/error-handling.md#wrap-errors).

```suggestion
		return fmt.Errorf("error getting VPC associated with Route53 Resolver DNSSEC config (%s): %w", d.Id(), err)
```

File: aws/resource_aws_route53_resolver_dnssec_config.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	// GetResolverDnssecConfig returns AccessDeniedException if sending a request with non-existing VPC id
+	if vpc == nil {
+		log.Printf("[WARN] VPC associated with Resolver DNSSEC config (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	raw, state, err := route53ResolverDnssecConfigRefresh(conn, d.Id())()
+	if err != nil {
+		return fmt.Errorf("error getting Route53 Resolver DNSSEC config (%s): %s", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_dnssec_config.go:101]: [Prefer `%w`](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/error-handling.md#wrap-errors).

```suggestion
		return fmt.Errorf("error getting Route53 Resolver DNSSEC config (%s): %w", d.Id(), err)
```

File: aws/resource_aws_route53_resolver_dnssec_config.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	log.Printf("[DEBUG] Deleting Route53 Resolver DNSSEC config: %s", d.Id())
+	_, err := conn.UpdateResolverDnssecConfig(&route53resolver.UpdateResolverDnssecConfigInput{
+		ResourceId: aws.String(d.Id()),
+		Validation: aws.String(route53resolver.ValidationDisable),
+	})
+	if isAWSErr(err, route53resolver.ErrCodeResourceNotFoundException, "") {
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("error deleting Route53 Resolver DNSSEC config (%s): %s", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_dnssec_config.go:131]: [Prefer `%w`](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/error-handling.md#wrap-errors).

```suggestion
		return fmt.Errorf("error deleting Route53 Resolver DNSSEC config (%s): %w", d.Id(), err)
```

File: aws/resource_aws_route53_resolver_dnssec_config.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	stateConf := &resource.StateChangeConf{
+		Pending:    pending,
+		Target:     target,
+		Refresh:    route53ResolverDnssecConfigRefresh(conn, id),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 5 * time.Second,
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("error waiting for Route53 Resolver DNSSEC config (%s) to reach target state: %s", id, err)
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_dnssec_config.go:154]:
```suggestion
		return fmt.Errorf("error waiting for Route53 Resolver DNSSEC config (%s) to reach target state: %w", id, err)
```

**Comment by ewbankkit** [comment]
@shuheiktgw Thanks for the contribution 👏. Overall it looks great.
I left a few nits on wrapping errors (I think the [Error Handling](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/error-handling.md) guide was only merged yesterday 😄).

DNSSEC configs [seem to have ARNs](https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonroute53resolver.html#amazonroute53resolver-resources-for-iam-policies) even though it's not returned in the API. Could you please add a Computed `arn` attribute (#13624), ensuring that the correct account ID is used: #16978? Thanks.

I need to think about the cross-service call to verify the VPC's existence. I see why it is being used but we usually try and avoid such calls.

File: aws/resource_aws_route53_resolver_dnssec_config.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+		Validation: aws.String(route53resolver.ValidationEnable),
+	}
+
+	log.Printf("[DEBUG] Creating Route53 Resolver DNSSEC config: %#v", req)
+	resp, err := conn.UpdateResolverDnssecConfig(req)
+	if err != nil {
+		return fmt.Errorf("error creating Route53 Resolver DNSSEC config: %s", err)
+	}
+
+	d.SetId(aws.StringValue(resp.ResolverDNSSECConfig.ResourceId))
```
Comment (ewbankkit) [inline aws/resource_aws_route53_resolver_dnssec_config.go:71]: Use `ResolvedDNSSECConfig.Id` as that is the actual resource's ID.
Agree that we should use the resource ID on import as that is what users would expect (as this is surfaced per-VPC in the AWS Console).

**Comment by ewbankkit** [comment]
It seems that you can update a DNSSEC config even if the associated resource (VPC) has been deleted:

```console
$ aws --region us-west-2 route53resolver list-resolver-dnssec-configs
{
    "ResolverDnssecConfigs": [
        {
            "Id": "rdsc-5a4960f0b41d3a5b",
            "OwnerId": "123456789012",
            "ResourceId": "vpc-0f2f5c603499f6c6b",
            "ValidationStatus": "DISABLED"
        },
        {
            "Id": "rdsc-9da58af749336ba",
            "OwnerId": "123456789012",
            "ResourceId": "vpc-0674233a9456aba43",
            "ValidationStatus": "DISABLED"
        },
        {
            "Id": "rdsc-be1866ecc1683e95",
            "OwnerId": "123456789012",
            "ResourceId": "vpc-0f993cbbf558be638",
            "ValidationStatus": "ENABLED"
        },
        {
            "Id": "rdsc-ca0760426d4a326b",
            "OwnerId": "123456789012",
            "ResourceId": "vpc-09925a270fdbb4206",
            "ValidationStatus": "ENABLED"
        }
    ]
}
$ aws --region us-west-2 route53resolver update-resolver-dnssec-config --resource-id "vpc-0f993cbbf558be638" --validation DISABLE
{
    "ResolverDNSSECConfig": {
        "Id": "rdsc-be1866ecc1683e95",
        "OwnerId": "123456789012",
        "ResourceId": "vpc-0f993cbbf558be638",
        "ValidationStatus": "DISABLING"
    }
}
```

**Comment by shuheiktgw** [comment]
@ewbankkit Thank you for your review! I believe I fixed them up so would you review the PR again? 🙏

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSRoute53ResolverDnssecConfig_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverDnssecConfig_ -timeout 120m
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_basic
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_basic
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_disappear
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_disappear
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_basic
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_disappear
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_disappear (397.50s)
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_basic (399.44s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	399.527s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @shuheiktgw!

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSRoute53ResolverDnssecConfig_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverDnssecConfig_ -timeout 120m
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_basic
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_basic
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_disappear
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_disappear
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_basic
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_disappear
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_disappear (380.47s)
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_basic (393.46s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	395.145s
```

Verified Acceptance Tests in GovCloud (us-gov-west-1)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSRoute53ResolverDnssecConfig_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverDnssecConfig_ -timeout 120m
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_basic
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_basic
=== RUN   TestAccAWSRoute53ResolverDnssecConfig_disappear
=== PAUSE TestAccAWSRoute53ResolverDnssecConfig_disappear
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_basic
=== CONT  TestAccAWSRoute53ResolverDnssecConfig_disappear
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_disappear (121.67s)
--- PASS: TestAccAWSRoute53ResolverDnssecConfig_basic (123.07s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	124.742s
```

**Comment by shuheiktgw** [comment]
Thank you all for your review!

**Pull request merged**